### DISCUSSION TITLE
Various issues when using entity.assign() with entities with collections

### DISCUSSION THREAD

**Discussion by mcpower** [mikro-orm/mikro-orm#3000] [Q&A]
**Describe the bug**
`wrap(entity).assign(newEntity)` does not work as expected:

1. Assigning another entity to an entity causes inserts - not updates - when flushed.
  The above, as-is, makes the entity manager "forget" about `entity`, and causes an insert instead of an update (resulting in primary key constraints failing).  
  To mitigate this issue, use `wrap(entity).assign(wrap(newEntity).toPOJO())`.
2. `toJSON()` and `toObject()` both return `undefined` where child entities would be in a one to many relationship.
  Using `wrap(entity).assign(wrap(newEntity).toJSON())` or `wrap(entity).assign(wrap(newEntity).toObject())` - assuming that `newEntity` has a one-to-many relation with items in them - causes an "invalid collection values" error due to `undefined` being in the collection. All items turn to "undefined", which goes against the TypeScript type definition.
  This is probably unrelated to `.assign`, but it is worth bringing up as `toJSON` and `toObject` might be used to mitigate issue 1.
3. Using assign on an entity with unpopulated collections results in those collections not being cleared when flushed.
  Mitigating the above using `toPOJO`, there's another issue: If the entity being assigned has unpopulated collections, and those collections are then assigned to, the collection in the database isn't *replaced* with the new collection - it is instead *appended to*.  
  To mitigate this issue, populate the entity being updated using `await em.populate(entity, true)` before assigning it.

**To Reproduce**
```ts
import {
  MikroORM,
  Entity,
  PrimaryKey,
  Options,
  Cascade,
  Collection,
  IdentifiedReference,
  ManyToOne,
  OneToMany,
  Property,
  Reference,
  wrap,
} from "@mikro-orm/core";
import { TsMorphMetadataProvider } from "@mikro-orm/reflection";
import type { SqliteDriver } from "@mikro-orm/sqlite";
import assert from "assert";

@Entity()
export class Author {
  @PrimaryKey()
  id: string;

  @OneToMany({
    entity: () => Book,
    mappedBy: "author",
    cascade: [Cascade.ALL],
    orphanRemoval: true,
  })
  books = new Collection<Book>(this);

  @Property()
  name: string;

  constructor(id: string, name: string) {
    this.id = id;
    this.name = name;
  }
}

@Entity()
export class Book {
  @PrimaryKey({ autoincrement: true })
  id?: number;

  @ManyToOne({ onDelete: "cascade" })
  author: IdentifiedReference<Author>;

  @Property()
  title: string;

  constructor(author: Author, title: string) {
    this.author = Reference.create(author);
    this.title = title;
  }
}

const config: Options<SqliteDriver> = {
  metadataProvider: TsMorphMetadataProvider,
  entities: [Author, Book],
  dbName: ":memory:",
  type: "sqlite",
  debug: true,
};

async function run() {
  const orm = await MikroORM.init<SqliteDriver>(config);
  // populate database
  {
    const em = orm.em.fork();
    await orm.getSchemaGenerator().createSchema();
    const jane = new Author("jsmith", "Jane Smith");
    jane.books.add(new Book(jane, "Book 1"));
    jane.books.add(new Book(jane, "Book 2"));
    await em.persistAndFlush(jane);
  }

  // update jsmith
  try {
    const em = orm.em.fork();
    const jen = new Author("jsmith", "Jen Smith");
    jen.books.add(new Book(jen, "Book 3"));
    jen.books.add(new Book(jen, "Book 4"));

    // note the lack of "populate" here
    const jsmith = await em.findOne(Author, "jsmith");
    // in a real application, this should be:
    // if (jsmith === null) {jsmith = jen; em.persist(jsmith);} else {...}
    assert(jsmith);
    wrap(jsmith).assign(jen); // BUG LIES HERE
    await em.flush();

    const em2 = orm.em.fork();
    // expected: Jen Smith with two books: Book 3 and Book 4
    console.log(await em2.findOne(Author, "jsmith", { populate: true }));
  } catch (err: unknown) {
    console.log(err);
  }

  await orm.close();
}

run();
```

Issue 1 is demonstrated above, resulting in this error:
```
UniqueConstraintViolationException: insert into `author` (`id`, `name`) values ('jsmith', 'Jen Smith') - SQLITE_CONSTRAINT_PRIMARYKEY: UNIQUE constraint failed: author.id
```

Issue 2 can be observed by replacing `wrap(jsmith).assign(jen);` with `wrap(jsmith).assign(wrap(jen).toJSON());` (or `toObject`), resulting in this error:
```
Error: Invalid collection values provided for 'Author.books' in Author.assign(): [ undefined, undefined ]
```

Issue 3 can be observed by replacing `wrap(jsmith).assign(jen);` with `wrap(jsmith).assign(wrap(jen).toPOJO());`, resulting in this output:
```
Author {
  id: 'jsmith',
  name: 'Jen Smith',
  books: Collection<Book> {
    '0': Book { id: 1, title: 'Book 1', author: [Reference] },
    '1': Book { id: 2, title: 'Book 2', author: [Reference] },
    '2': Book { id: 3, title: 'Book 3', author: [Reference] },
    '3': Book { id: 4, title: 'Book 4', author: [Reference] },
    initialized: true,
    dirty: false
  }
}
```
Book 1 and Book 2 should not be there.

**Expected behavior**
All issues can be resolved by replacing `wrap(jsmith).assign(jen);` with
```ts
await em.populate(jsmith, true);
wrap(jsmith).assign(wrap(jen).toPOJO());
```
which results in the expected output:
```
Author {
  id: 'jsmith',
  name: 'Jen Smith',
  books: Collection<Book> {
    '0': Book { id: 3, title: 'Book 3', author: [Reference] },
    '1': Book { id: 4, title: 'Book 4', author: [Reference] },
    initialized: true,
    dirty: false
  }
}
```

**Additional context**
Feel free to split this into multiple issues.

**Versions**

```json
"dependencies": {
  "@mikro-orm/core": "^5.1.2",
  "@mikro-orm/reflection": "^5.1.2",
  "@mikro-orm/sqlite": "^5.1.2",
  "ts-node": "^10.7.0",
  "typescript": "^4.6.3"
}
```

**Comment by B4nan** [comment]
You are not supposed to be putting an entity instance there, it needs to be entity DTO - a POJO - without a prototype. If you pass an entity instance there, it means you want to _replace_ what is there, not assign/merge - an entity instance is considered as its PK. And you are not supposed to be passing entity instance instead of the whole payload, it can be a property value (e.g. `author: new Author()`), but not the whole payload.

> Error: Invalid collection values provided for 'Author.books' in Author.assign(): [ undefined, undefined ]

This is because you are serializing unflushed entity, which on its own is a misusage. You end up doing `assign({ books: [undefined, undefined] })`, hence the validation you are getting.

Correct usage of `assign` is something like this. Note that I am passing a POJO into data.

```ts
wrap(jsmith).assign({
    id: 'jsmith',
    name: "Jen Smith",
    books: [
        { name: "Book 3" },
        { name: "Book 4" },
    ],
});
```

**Reply by mcpower** [reply]
Sorry, X/Y problem. The specific thing I'm trying to do is "upsert" a given unflushed entity instance. My business logic creates unflushed entities, which I usually insert into the database using a `persist`. Is it possible to instead update the database with an unflushed entity, given that the primary key is already in the database?

See https://github.com/mcpower/discord-link-organiser/blob/219b526ebf4e1f77fe516739cae13b1e30784a9b/src/discord/client.ts#L196 for the specific code I'm encountering this issue in.

**Reply by B4nan** [reply]
You should never use the entity constructor for managed entities. To construct managed entity from DTO, you can use `em.merge(Author, { ... })` which works the same as if you would query the database for that entity (but without a query, using your provided data instead). Alternatively you can as well use `em.map()` which also handles conversion of column names to property names (e.g. in case your data contains things like `first_name` but the property is called `firstName`). You should also do all of this in an explicit transaction. 

Something like this could work:

```ts
const data = {
    id: 'jsmith',
    name: "Jen Smith",
    books: [
        { name: "Book 3" },
        { name: "Book 4" },
    ],
};

// things will get flushed automatically with `em.transactional`
const author = await em.transactional(async () => {
    let exists = await em.findOne(Author, data.id);

    if (exists) {
        // optionally we might want to update the existing record
        wrap(exists).assign(data);
    } else {
        exists = em.create(Author, data, { persist: true });
    }

    return exists;
});
```

Or another alternative could be using `insert` with `on conflict`, which would require a QB. Not sure if SQLite supports that though.

**Reply by mcpower** [reply]
My business logic currently creates unmanaged entities. If I'm understanding you correctly, is your recommendation to make the business logic return POJOs, instead of unmanaged entities?

To be very specific:

- I have a managed entity gotten by a `em.findOne`.
- I want to update it given an object from my business logic, so the ORM only flushes changed properties (including collections).

If I don't change my business logic (and it still returns unmanaged entities), the question I am asking is: What is the best way to update a managed entity using an unmanaged entity (with the same PK)? At the moment, that is:
```ts
// Populate the entity to remove existing entities in collections.
// Can this be done automatically without populating with a SQL query like:
//   delete from `author` as `a0` where `a0`.id = 'jsmith'
// ?
await em.populate(jsmith, true);
wrap(jsmith).assign(wrap(jen).toPOJO());
```

If I do change my business logic to return POJOs instead and I use your code, it gives me this error:
```
[info] MikroORM version: 5.1.2
...
[query] begin
[query] select `a0`.* from `author` as `a0` where `a0`.`id` = 'jsmith' limit 1 [took 60003 ms]
[query] rollback
DriverException: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?
previous KnexTimeoutError: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?
  sql: "select `a0`.* from `author` as `a0` where `a0`.`id` = 'jsmith' limit 1",
  bindings: undefined
}
```

I replaced the try/except in the code above with:
```ts
  // update jsmith
  try {
    const em = orm.em.fork();
    const data = {
      id: "jsmith",
      name: "Jen Smith",
      books: [{ title: "Book 3" }, { title: "Book 4" }],
    };

    // things will get flushed automatically with `em.transactional`
    const author = await em.transactional(async () => {
      let exists = await em.findOne(Author, data.id);

      if (exists) {
        // optionally we might want to update the existing record
        wrap(exists).assign(data);
      } else {
        exists = em.create(Author, data, { persist: true });
      }

      return exists;
    });

    const em2 = orm.em.fork();
    // expected: Jen Smith with two books: Book 3 and Book 4
    console.log(await em2.findOne(Author, "jsmith", { populate: true }));
  } catch (err: unknown) {
    console.log(err);
  }
```

Removing the `em.transactional` gives me this error instead:

```
TypeError: Cannot read properties of undefined (reading '__meta')
```

Specifically, this code:

```ts
  // update jsmith
  try {
    const em = orm.em.fork();
    const data = {
      id: "jsmith",
      name: "Jen Smith",
      books: [{ title: "Book 3" }, { title: "Book 4" }],
    };

    let exists = await em.findOne(Author, data.id);

    if (exists) {
      // optionally we might want to update the existing record
      wrap(exists).assign(data);
    } else {
      exists = em.create(Author, data, { persist: true });
    }
    await em.flush();

    const em2 = orm.em.fork();
    // expected: Jen Smith with two books: Book 3 and Book 4
    console.log(await em2.findOne(Author, "jsmith", { populate: true }));
  } catch (err: unknown) {
    console.log(err);
  }
```
